95.md

```diff
@@ -0,0 +1,91 @@
+NIP-95
+======
+
+File Storage
+------------
+
+`draft` `optional` `author:arthurfranca` `author:Semisol` `author:frbitten` `author:staab`
+
+## Motivation
+
+This NIP defines a way to turn a relay into a storage service.
+The spec doesn't use regular nostr events through websockets for
+storing, requesting nor retrieving data because it would be inneficient.
+
+## Relay Adaptation
+
+Relays wishing to be used as file storage should opt-in by making available
+an https route at `/nip95(/<sha256-file-hash>)` for handling files as described below.
+They may announce support with [NIP-11](11.md) `supported_nips` field.
+Other ways of advertising NIP-95 support may be described later.
+
+## Upload
+
+A file can be uploaded one at a time to `https://relay.domain/nip95` as `multipart/form-data` content type using `POST` method with the file object set to the `file` form data field.
+
+Clients must add a [NIP-98](98.md) `Authorization` header with the encoded `payload` tag set to the base64-encoded 256-bit SHA-256 hash of the file (not the hash of the whole request body).
+If using an html form, use an `Authorization` form data field instead.
+
+Others custom form data fields may be added depending on specific relay support.
+
+The `filename` embedded in the file may not be honored by the relay, which can internally store just the filename as the SHA-256 hash value ignoring extra metadata.
+
+Relays must reject the upload if the `payload` tag value contains an already used SHA-256 hash or
+it isn't the same of the received file.
+
+The upload response is a json as follows:
+
+```js
+{
+  nip95: {
+    // SHA-256 hash. Empty string if unsuccessful
+    x: "719171db19525d9d08dd69cb716a18158a249b7b3b3ec4bbdec5698dca104b7b"
+  },
+  errors: {
+    // Empty array if successful
+    nip95: ['error 1', 'error 2']
+  }
+}
+```
+
+Clients may upload the same file to one or many relays.
+After successful upload, the client may optionally generate and send to any set of relays an [NIP-94](94.md) event with atleast these tags:
+
+```js
+[
+  ["url", "https://relay.domain/nip95/719171db19525d9d08dd69cb716a18158a249b7b3b3ec4bbdec5698dca104b7b"],
+  ["x", "719171db19525d9d08dd69cb716a18158a249b7b3b3ec4bbdec5698dca104b7b"]
+]
+```
+
+If the file upload was made to more than one relay, `r` query params must be added to the url string, containing each of the relays used in successful upload operations. For example:
+
+```js
+[
+  ["url", "https://relay.domain/nip95/<sha256-file-hash>?r=relay.domain2&r=relay.domain3&r=relay.domain4"],
+  ["x", <the-hash>]
+]
+```
+
+Alternativelly, instead of using NIP-94, one can share or embbed on a note just the above url, optionally with added [NIP-54](54.md) inline metadata.
```

Comment:
I prefer the approach I mentioned in https://github.com/nostr-protocol/nips/pull/345#issuecomment-1538640440, where a new event kind (or an adapted NIP 94 event) gets published. This can be replicated/re-posted, and if relays want to store the pointer event, they are also required to host the file itself. This allows the file replication to grow/shrink dynamically instead of be limited to the user's original relay selection. It also fixes the redirect problem below, because the file is guaranteed to exist at the same relay the pointer event was found on.

Reply:
On your approach the relay creates the NIP-94-like event and reference the uploaded file by the event id. Each relay that receives the file will be referenced by a different event id although it is the same file.

For relays to cooperate (issue a 302 redirect when they don't have the file anymore), the file must be referenced by the sha256 hash. This is probably the best feature of this NIP imo. The infinite redirect won't be a problem if limiting the number of `r` query params when redirecting.

When the job to create the NIP-94 is left to the user's client, the file storage relay don't need to be a regular nostr relay. Don't even need websockets. It just requires NIP-98 HTTP Auth support.

Reply:
That's true, but I wonder how relevant that is? Sounds like more of a use-case for torrents, since more references support better seeding. I would imagine a low level of duplication of content for the nostr use case, where the value of the file is related to the publisher's pubkey and context, rather than the content itself (e.g., memes vs movies).